File: src/main/java/me/chriss99/InputDeviceManager.java
package me.chriss99;

import static org.lwjgl.glfw.GLFW.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.function.BiConsumer;

public class InputDeviceManager {
    private final HashMap<Integer, ArrayList<Runnable>> keyPressRunnables = new HashMap<>();
    private final HashMap<Integer, ArrayList<Runnable>> keyReleaseRunnables = new HashMap<>();
    private final ArrayList<BiConsumer<Double, Double>> mouseAbsoluteMovementConsumers = new ArrayList<>();
    private final ArrayList<BiConsumer<Double, Double>> mouseRelativeMovementConsumers = new ArrayList<>();
    private final ArrayList<BiConsumer<Double, Double>> mouseScrollConsumers = new ArrayList<>();

    private double lastMouseX = Double.NaN;
    private double lastMouseY = Double.NaN;

    public InputDeviceManager(long windowId) {
        glfwSetKeyCallback(windowId, (window, key, scancode, action, mods) -> {
            ArrayList<Runnable> runnables = new ArrayList<>();
            switch (action) {
                case GLFW_PRESS -> runnables = keyPressRunnables.get(key);
                case GLFW_RELEASE -> runnables = keyReleaseRunnables.get(key);
            }

            if (runnables == null)
                return;

            for (Runnable runnable : runnables)
                runnable.run();
        });

        glfwSetCursorPosCallback(windowId, (win, x, y) -> {
            for (BiConsumer<Double, Double> consumer : mouseAbsoluteMovementConsumers)
                consumer.accept(x, y);

            if (Double.isNaN(lastMouseX)) {
                lastMouseX = x;
                lastMouseY = y;
                return;
            }

            double dx = x - lastMouseX;
            double dy = y - lastMouseY;
            for (BiConsumer<Double, Double> consumer : mouseRelativeMovementConsumers)
                consumer.accept(dx, dy);

            lastMouseX = x;
            lastMouseY = y;
        });

        glfwSetScrollCallback(windowId, (win, dx, dy) -> {
            for (BiConsumer<Double, Double> consumer : mouseScrollConsumers)
                consumer.accept(dx, dy);
        });
    }


    public void addKeyPressRunnable(int key, Runnable runnable) {
        keyPressRunnables.computeIfAbsent(key, k -> new ArrayList<>()).add(runnable);
    }

    public void addKeyReleaseRunnable(int key, Runnable runnable) {
        keyReleaseRunnables.computeIfAbsent(key, k -> new ArrayList<>()).add(runnable);
    }

    public void addMouseAbsoluteMovementConsumer(BiConsumer<Double, Double> consumer) {
        mouseAbsoluteMovementConsumers.add(consumer);
    }

    public void addMouseRelativeMovementConsumer(BiConsumer<Double, Double> consumer) {
        mouseRelativeMovementConsumers.add(consumer);
    }

    public void addMouseScrollConsumer(BiConsumer<Double, Double> consumer) {
        mouseScrollConsumers.add(consumer);
    }

    public boolean removeKeyPressRunnable(int key, Runnable runnable) {
        return keyPressRunnables.computeIfAbsent(key, k -> new ArrayList<>()).add(runnable);
    }

    public boolean removeKeyReleaseRunnable(int key, Runnable runnable) {
        return keyReleaseRunnables.computeIfAbsent(key, k -> new ArrayList<>()).remove(runnable);
    }

    public boolean removeMouseAbsoluteMovementConsumer(BiConsumer<Double, Double> consumer) {
        return mouseAbsoluteMovementConsumers.add(consumer);
    }

    public boolean removeMouseRelativeMovementConsumer(BiConsumer<Double, Double> consumer) {
        return mouseRelativeMovementConsumers.add(consumer);
    }

    public boolean removeMouseScrollConsumer(BiConsumer<Double, Double> consumer) {
        return mouseScrollConsumers.add(consumer);
    }
}

File: src/main/kotlin/InputDeviceManager.kt
package me.chriss99

import glm_.vec2.Vec2d
import org.lwjgl.glfw.GLFW

class InputDeviceManager(windowId: Long) {
    private val keyPressCallbacks = HashMap<Int, ArrayList<() -> Unit>>()
    private val keyReleaseCallbacks = HashMap<Int, ArrayList<() -> Unit>>()
    private val mouseAbsoluteMovementConsumers = ArrayList<(pos: Vec2d) -> Unit>()
    private val mouseRelativeMovementConsumers = ArrayList<(change: Vec2d) -> Unit>()
    private val mouseScrollConsumers = ArrayList<(change: Vec2d) -> Unit>()

    private var lastMousePos = Vec2d(Double.NaN)

    init {
        GLFW.glfwSetKeyCallback(windowId) { _, key: Int, _, action: Int, _ ->
            val callbacks: ArrayList<() -> Unit>? = when (action) {
                GLFW.GLFW_PRESS -> keyPressCallbacks[key]
                GLFW.GLFW_RELEASE -> keyReleaseCallbacks[key]
                else -> null
            }

            callbacks?.forEach { it() }
        }

        GLFW.glfwSetCursorPosCallback(windowId) { _, x: Double, y: Double ->
            val pos = Vec2d(x, y)
            mouseAbsoluteMovementConsumers.forEach { it(pos) }

            if (java.lang.Double.isNaN(lastMousePos.x)) {
                lastMousePos = pos
                return@glfwSetCursorPosCallback
            }

            val change = pos - lastMousePos
            mouseRelativeMovementConsumers.forEach { it(change) }

            lastMousePos = pos
        }

        GLFW.glfwSetScrollCallback(windowId) { _, dx: Double, dy: Double ->
            val change = Vec2d(dx, dy)
            mouseScrollConsumers.forEach { it(change) }
        }
    }


    fun addKeyPressCallback(key: Int, runnable: () -> Unit) {
        keyPressCallbacks.computeIfAbsent(key) { _ -> ArrayList() }.add(runnable)
    }

    fun addKeyReleaseCallback(key: Int, runnable: () -> Unit) {
        keyReleaseCallbacks.computeIfAbsent(key) { _ -> ArrayList() }.add(runnable)
    }

    fun addMouseAbsoluteMovementConsumer(consumer: (pos: Vec2d) -> Unit) {
        mouseAbsoluteMovementConsumers.add(consumer)
    }

    fun addMouseRelativeMovementConsumer(consumer: (change: Vec2d) -> Unit) {
        mouseRelativeMovementConsumers.add(consumer)
    }

    fun addMouseScrollConsumer(consumer: (change: Vec2d) -> Unit) {
        mouseScrollConsumers.add(consumer)
    }

    fun removeKeyPressCallback(key: Int, runnable: () -> Unit): Boolean {
        return keyPressCallbacks.computeIfAbsent(key) { _ -> ArrayList() }.remove(runnable)
    }

    fun removeKeyReleaseCallback(key: Int, runnable: () -> Unit): Boolean {
        return keyReleaseCallbacks.computeIfAbsent(key) { _ -> ArrayList() }.remove(runnable)
    }

    fun removeMouseAbsoluteMovementConsumer(consumer: (pos: Vec2d) -> Unit): Boolean {
        return mouseAbsoluteMovementConsumers.remove(consumer)
    }

    fun removeMouseRelativeMovementConsumer(consumer: (change: Vec2d) -> Unit): Boolean {
        return mouseRelativeMovementConsumers.remove(consumer)
    }

    fun removeMouseScrollConsumer(consumer: (change: Vec2d) -> Unit): Boolean {
        return mouseScrollConsumers.remove(consumer)
    }
}